src/cocotb/clock.py

```diff
@@ -152,42 +186,84 @@ async def start(self, start_high: bool = True) -> None:
 
-        t_high = self.period // 2
+        period = get_sim_steps(self._period, self._units)
+        t_high = period // 2
 
-        if self.impl == "gpi":
-            clkobj = clock_create(self.signal._handle)
-            clkobj.start(self.period, t_high, start_high)
+        if self._impl == "gpi":
+            self._clkobj = clock_create(self._signal._handle)
+            self._clkobj.start(period, t_high, start_high)
 
-            try:
+            async def drive() -> None:
                 # The clock is meant to toggle forever, so awaiting this should
-                # never return (except in case of CancelledError).
-                # Await on an event that's never set.
+                # never return by awaiting on Event that's never set.
                 e = Event()
                 await e.wait()
-            finally:
-                clkobj.stop()
+
         else:
-            timer_high = Timer(t_high)
-            timer_low = Timer(self.period - t_high)
-            if start_high:
-                self.signal.set(1)
-                await timer_high
-            while True:
-                self.signal.set(0)
-                await timer_low
-                self.signal.set(1)
-                await timer_high
 
-    def __str__(self) -> str:
-        return type(self).__qualname__ + f"({self.frequency:3.1f} MHz)"
+            async def drive() -> None:
+                timer_high = Timer(t_high)
+                timer_low = Timer(period - t_high)
+                if start_high:
+                    self._signal.set(1)
+                    await timer_high
+                while True:
+                    self._signal.set(0)
+                    await timer_low
+                    self._signal.set(1)
+                    await timer_high
+
+        self._task = cocotb.start_soon(drive())
+
+        # So if a user calls `task.kill()` on the returned task the Clock object is stopped.
+        self._task._add_done_callback(lambda _: self._cleanup())
+
+        return self._task
+
+    def stop(self) -> None:
+        """Stop driving the clock signal.
+
+        .. versionadded:: 2.0
+        """
+        if self._task is None:
+            raise RuntimeError("Stopping clock that was never started.")
+        self._task.kill()
+        self._cleanup()
+
+    def _cleanup(self) -> None:
+        if self._impl == "gpi":
+            self._clkobj.stop()
+        self._task = None
+
+    async def cycles(
+        self,
+        num_cycles: int,
+        edge_type: Union[Type[RisingEdge], Type[FallingEdge], Type[Edge]] = RisingEdge,
+    ) -> None:
+        """Wait for a number of clock cycles."""
+        # TODO Improve implementation to use a Timer to skip most of the cycles
+        await ClockCycles(self._signal, num_cycles, edge_type)
+
+    def __repr__(self) -> str:
+        return f"{type(self).__qualname__}({self._signal._path}, {self._period}, {self._units!r})"
```

Comment:
Maybe it would be convenient to also get the frequency that results from the period here.

Reply:
Would implementing `__str__` be helpful for users? IIRC this is meant to be more abstract.

Reply:
I don't think so. `__repr__` was definitely what was meant here. I'll go back to the original impl.